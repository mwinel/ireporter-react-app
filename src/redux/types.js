export const signupTypes = {
  SIGNUP_SUCCESS: "SIGNUP_SUCCESS"
};

export const loginTypes = {
  LOGIN_SUCCESS: "LOGIN_SUCCESS"
};
