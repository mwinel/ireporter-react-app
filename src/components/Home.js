import React, { Component } from "react";
import Navbar from '../components/Layout/Navbar';
import Footer from '../components/Layout/Footer';

class Home extends Component {
  render () {
    return (
      <div>
        <Navbar />
        <div className="container">
          <div className="row">
            <div className="flag-content">
              <h1>Welcome to iReporter!</h1>
            </div>
          </div>
          <Footer />
        </div>
      </div>
    );
  }
}

export default Home;
