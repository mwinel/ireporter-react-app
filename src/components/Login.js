import React, { Component } from 'react';
import { connect } from 'react-redux';
import Navbar from './Layout/Navbar';
import Footer from './Layout/Footer';
import login from '../redux/actions/loginAction';

export class Login extends Component {
  constructor (props) {
    super(props);
    this.state = {
      username: '',
      password: ''
    };
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  componentWillMount () {
    this.props.login();
  }

  onChange (e) {
    this.setState({ [e.target.name]: e.target.value });
  }

  onSubmit (e) {
    e.preventDefault();
    const loginUser = {
      username: this.state.username,
      password: this.state.password
    };
    this.props.login(loginUser);
  }

  render () {
    return (
      <div>
        <Navbar />
        <div className="row">
          <form id="userLogin" onSubmit={this.onSubmit}>
            <h2>Login to iReporter</h2>
            <input type="text" name="username" onChange={this.onChange} value={this.state.username} placeholder="Enter username" />
            <input type="text" name="password" onChange={this.onChange} value={this.state.password} placeholder="Enter password" />
            <button type="submit" className="btn-login">Login</button>
            <p className="form-footer">Don't have an account <a href="./signup">Sign Up</a>.</p>
          </form>
        </div>
        <Footer />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  login: state.login.login
});

export default connect(mapStateToProps, { login })(Login);
