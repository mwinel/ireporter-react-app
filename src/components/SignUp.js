import React, { Component } from 'react';
import { connect } from 'react-redux';
import Navbar from '../components/Layout/Navbar';
import Footer from '../components/Layout/Footer';
import signUp from '../redux/actions/signUpAction';

export class SignUp extends Component {
  constructor (props) {
    super(props);
    this.state = {
      firstname: '',
      lastname: '',
      othernames: '',
      username: '',
      email: '',
      password: '',
      phone_number: ''
    };
    this.onChange = this.onChange.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
  }

  componentWillMount () {
    this.props.signUp();
  }

  onChange (e) {
    this.setState({ [e.target.name]: e.target.value });
  }

  onSubmit (e) {
    e.preventDefault();
    const signup = {
      firstname: this.state.firstname,
      lastname: this.state.lastname,
      othernames: this.state.othernames,
      username: this.state.username,
      email: this.state.email,
      password: this.state.password,
      phone_number: this.state.phone_number
    };
    this.props.signUp(signup);
  }

  render () {
    return (
      <div>
        <Navbar />
        <div className="row">
          <form id="addUser" onSubmit={this.onSubmit}>
            <h2>Please Sign Up</h2>
            <input type="text" name="firstname" onChange={this.onChange} value={this.state.firstname} placeholder="Enter firstname" />
            <input type="text" name="lastname" onChange={this.onChange} value={this.state.lastname} placeholder="Enter lastname" />
            <input type="text" name="othernames" onChange={this.onChange} value={this.state.othernames} placeholder="Enter othernames" />
            <input type="text" name="username" onChange={this.onChange} value={this.state.username} placeholder="Enter username" />
            <input type="text" name="email" onChange={this.onChange} value={this.state.email} placeholder="Enter email" />
            <input type="text" name="password" onChange={this.onChange} value={this.state.password} placeholder="Enter password" />
            <input type="text" name="phone_number" onChange={this.onChange} value={this.state.phone_number} placeholder="Enter phone number" />
            <button type="submit" className="btn-signup">Sign Up</button>
            <p className="form-footer">Already have an account <a href="./login">Login</a>.</p>
          </form>
        </div>
        <Footer />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  user: state.user.user
});

export default connect(mapStateToProps, { signUp })(SignUp);
