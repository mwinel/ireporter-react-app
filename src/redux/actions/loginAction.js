import { loginTypes } from '../types';
import { loginUrl } from '../../urls';

const login = loginData => dispatch => {
  return fetch(loginUrl, {
    method: 'POST',
    headers: {
      'content-type': 'application/json'
    },
    body: JSON.stringify(loginData)
  })
    .then(res => res.json())
    .then(data => {
      dispatch({
        type: loginTypes.LOGIN_SUCCESS,
        payload: data
      });
      if (data.message) {
        alert(`${data.message}`);
        window.setTimeout(function () {
          window.location.replace('/');
          window.location.reload(true);
        }, 1000);
      } else if (data.error) {
        alert(`${data.error}`);
      }
    });
};

export default login;
