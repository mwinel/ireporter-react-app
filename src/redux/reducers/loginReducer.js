import { loginTypes } from '../types';

const initialState = {
  login: {},
  errors: {}
};

export default function (state = initialState, action) {
  switch (action.type) {
  case loginTypes.LOGIN_SUCCESS:
    return {
      ...state,
      login: action.payload
    };
  default:
    return state;
  }
}
