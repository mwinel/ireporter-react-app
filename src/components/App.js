import React, { Component } from "react";
import { Provider } from 'react-redux';
import { BrowserRouter, Route } from 'react-router-dom';
import '../styles/App.css';
import store from '../redux/store';
import Home from '../components/Home';
import SignUp from '../components/SignUp';
import Login from '../components/Login';

class App extends Component {
  render () {
    return (
      <BrowserRouter>
        <Provider store={store}>
          <Route exact path="/" component={Home} />
          <Route exact path="/signup" component={SignUp} />
          <Route exact path="/login" component={Login} />
        </Provider>
      </BrowserRouter>
    );
  }
}

export default App;
