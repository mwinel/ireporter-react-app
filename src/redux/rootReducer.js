import { combineReducers } from 'redux';
import indexReducer from './reducers/indexReducer';
import signUpReducer from './reducers/signUpReducer';
import loginReducer from './reducers/loginReducer';

export default combineReducers({
  test: indexReducer,
  user: signUpReducer,
  login: loginReducer
});
