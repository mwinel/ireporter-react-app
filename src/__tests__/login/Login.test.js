import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import fetchMock from 'fetch-mock'
import loginReducer from '../../redux/reducers/loginReducer';
import loginAction from '../../redux/actions/loginAction';
import { Login } from '../../components/Login';
import { loginTypes } from '../../redux/types';

const storeItems = {};
const middlewares = [thunk]
const mockStore = configureStore(middlewares);
const store = mockStore(storeItems);

const initialState = {};

const props = {
  username: '',
  password: '',
  onChange: jest.fn(),
  onSubmit: jest.fn(),
  login: jest.fn(),
  preventDefault: jest.fn()
}

const mockEvent = {
  preventDefault: jest.fn()
}

describe('Login Component', () => {
  let LoginComponent = shallow(<Login {...props} store={store}/>)
  it('Should match snapshot', () => {
    LoginComponent.instance().onSubmit(mockEvent);
    expect(toJson(LoginComponent)).toMatchSnapshot();
  });

  it('should trigger onChange on login', () => {
    const login = { 
      username: 'test',
      password: 'test'
    }
    const evt = {
      preventDefault: jest.fn(),
      target: login
    }
    expect(LoginComponent.instance().onChange(evt));
  });
});

describe("login success", () => {
  it('tests login success', () => {
    const newState = loginReducer(initialState,
      { type: loginTypes.LOGIN_SUCCESS });
    expect(newState).toEqual({});
  });
});

describe("login action", () => {
  afterEach(() => {
    fetchMock.restore()
  })

  it("creates LOGIN_SUCCESS when logging in a user", () => {
    fetchMock.post('*', {
      body: {
        username: 'mwinelnnnnn',
        password: 'password'
      },
      headers: { 'content-type': 'application/json' }
    })

    const expectedActions = [
      { type: loginTypes.LOGIN_SUCCESS, 
        payload: {
          username: 'mwinelnnnnn',
          password: 'password'
        }
      }
    ]
    return store.dispatch(loginAction()).then(() => {
      expect(store.getActions()).toEqual(expectedActions)
    })
  })
})
