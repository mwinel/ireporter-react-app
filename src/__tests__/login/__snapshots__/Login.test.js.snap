// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Component Should match snapshot 1`] = `
<div>
  <Navbar />
  <div
    className="row"
  >
    <form
      id="userLogin"
      onSubmit={[Function]}
    >
      <h2>
        Login to iReporter
      </h2>
      <input
        name="username"
        onChange={[Function]}
        placeholder="Enter username"
        type="text"
        value=""
      />
      <input
        name="password"
        onChange={[Function]}
        placeholder="Enter password"
        type="text"
        value=""
      />
      <button
        className="btn-login"
        type="submit"
      >
        Login
      </button>
      <p
        className="form-footer"
      >
        Don't have an account 
        <a
          href="./signup"
        >
          Sign Up
        </a>
        .
      </p>
    </form>
  </div>
  <Footer />
</div>
`;
